### DISCUSSION TITLE
Open SSL and TLS interoperability issue

### DISCUSSION THREAD

**Discussion by sa791017** [openssl/openssl#24596] [Q&A]
I have no communication issues with my ssl client and ssl server.
But when I replace the ssl client with mbedtls client, I receive an encryption alert and connection reset from the server for my application messages.

The client handshake and certificate verification works fine for both the clients. 

On the SSL server end , in the logs I see [error] <0.6112.7> {bad_header,<<0,0,0,0,0,0,0,0>>} 
This error message occurs only with mbedtls client.

I want to know what is the difference in the headers, for the TLS Record protocol and SSL Record protocol. 
My  ssl client and mbedtls client sent the same 8bytes of data
unsigned char buf1[8] = { 'A','M','Q','P',0,0,9,1 };
But I notice in the wireshark , that the ssl client packet length is 32 bytes while for the mbedtls its 29bytes.  
These additional data are part of ssl and mbedtls protocol header and encryption as below.

 
 

As per the below link, when there is no more data to send, the sender sends this TLS Alert.
https://ask.wireshark.org/question/2492/i-am-getting-a-encryption-alert-from-the-server-and-connection-resets/
My assumption is that the ssl server end is reading the data and sends the ACK followed by the TLS alert and then closing the connection.
So TLS client communication is working as expected. But the SSL server is not able to interpret the same data with mbedtls header as expected.

The TLS debug file from wire shark shows the below message.
dissect_ssl enter frame #303 (already visited)
packet_from_server: is from server - TRUE
  conversation = 000000000CC53A60, ssl_session = 0000000000000000
  record: offset = 0, reported_length_remaining = 43
dissect_ssl3_record: content_type 20 Change Cipher Spec
  record: offset = 6, reported_length_remaining = 37
dissect_ssl3_record: content_type 22 Handshake

dissect_ssl enter frame #305 (already visited)
packet_from_server: is from server - FALSE
  conversation = 000000000CC53A60, ssl_session = 0000000000000000
  record: offset = 0, reported_length_remaining = 29
dissect_ssl3_record: content_type 23 Application Data

dissect_ssl enter frame #307 (already visited)
packet_from_server: is from server - TRUE
  conversation = 000000000CC53A60, ssl_session = 0000000000000000
  record: offset = 0, reported_length_remaining = 29
dissect_ssl3_record: content_type 23 Application Data

dissect_ssl enter frame #309 (already visited)
packet_from_server: is from server - TRUE
  conversation = 000000000CC53A60, ssl_session = 0000000000000000
  record: offset = 0, reported_length_remaining = 23
dissect_ssl3_record: content_type 21 Alert

3.The SSL server, logs record the below error
2021-12-30 23:59:49.548000-05:00 [info] <0.6112.7> accepting AMQP connection <0.6112.7> ([::1]:56528 -> [::1]:5671)
2021-12-30 23:59:49.548000-05:00 [error] <0.6112.7> closing AMQP connection <0.6112.7> ([::1]:56528 -> [::1]:5671):
2021-12-30 23:59:49.548000-05:00 [error] <0.6112.7> {bad_header,<<0,0,0,0,0,0,0,0>>}

How can I fix this issue? Any inputs or suggestions?

Please help .

**Comment by ljluestc** [comment]
```

#include "mbedtls/ssl.h"
#include "mbedtls/net_sockets.h"
#include "mbedtls/debug.h"

// Initialize the context
mbedtls_ssl_context ssl;
mbedtls_ssl_config conf;

// Initialize the SSL/TLS structure
mbedtls_ssl_init(&ssl);
mbedtls_ssl_config_init(&conf);

// Set the default configuration
mbedtls_ssl_config_defaults(&conf,
                            MBEDTLS_SSL_IS_CLIENT,
                            MBEDTLS_SSL_TRANSPORT_STREAM,
                            MBEDTLS_SSL_PRESET_DEFAULT);

// Set debug level
mbedtls_debug_set_threshold(3);

// Set the desired SSL/TLS version
mbedtls_ssl_conf_min_version(&conf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);
mbedtls_ssl_conf_max_version(&conf, MBEDTLS_SSL_MAJOR_VERSION_3, MBEDTLS_SSL_MINOR_VERSION_3);

// Attach the configuration to the SSL context
mbedtls_ssl_setup(&ssl, &conf);

// Set the server hostname
mbedtls_ssl_set_hostname(&ssl, "your_server_hostname");

// Set up the bio callbacks
mbedtls_ssl_set_bio(&ssl, &server_fd, mbedtls_net_send, mbedtls_net_recv, NULL);

// Handshake
if ((ret = mbedtls_ssl_handshake(&ssl)) != 0) {
    // Handle error
}

// Prepare the data to be sent
unsigned char buf1[8] = { 'A','M','Q','P',0,0,9,1 };

// Send data
if ((ret = mbedtls_ssl_write(&ssl, buf1, sizeof(buf1))) <= 0) {
    // Handle error
}

// Clean up
mbedtls_ssl_close_notify(&ssl);
mbedtls_ssl_free(&ssl);
mbedtls_ssl_config_free​⬤


```
